src/main/java/org/manifold/compiler/front/ExpressionGraph.java

```diff
@@ -164,9 +170,8 @@ public void addSubExpressionGraph(ExpressionGraph subGraph,
     // map of subgraph edge -> new edge to be inserted
     Map<ExpressionEdge, ExpressionEdge> exprEdgeMap = new HashMap<>();
     subGraph.getEdges().forEach(e -> {
-        // copy source/target for now since ExpressionEdge doesn't like creating with null on both
-        // they will be replaced with the correct vertices later
-        ExpressionEdge newEdge = new ExpressionEdge(e.getSource(), e.getTarget());
+        // replace these with the correct vertices later
+        ExpressionEdge newEdge = new ExpressionEdge(null, null);
```

Comment:
I don't have a short-term fix here, but in general I think we can get around these gross intermediate states by lazily evaluating things and allowing trees to exist ephemerally in the call stack rather than as a literal data structure.